pkg/apis/openebs.io/scp/v1alpha1/storage-cohort.go

```diff
@@ -27,10 +28,129 @@ import (
 // +kubebuilder:object:root=true
 // +kubebuilder:resource:scope=Namespaced
 type StorageCohort struct {
-	metav1.TypeMeta   `json:",inline"`
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object's metadata.
 	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the behavior of a cohort.
+	// +required
+	Spec StorageCohortSpec `json:"spec"`
+
+	// Most recently observed status of the cohort.
+	// Populated by the cohort operator or cohort manager.
+	// +optional
+	Status StorageCohortStatus `json:"status,omitempty"`
 }
 
+// StorageCohortSpec describes the attributes that a cohort is created with.
+type StorageCohortSpec struct {
+	// NodeSelector is used to specify the cohort to be considered
+	// +optional
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+	// CohortManager stores all the details about the cohort manager responsible for
+	// managing the cohort
+	// +optional
+	CohortManager interface{} `json:"cohortManager,omitempty"`
+
+	// DefaultStorageProvisioner is the default provisioner for the cohort which can be used
+	// for provisioning pools or volumes when no provisioner is specified in the storage pool
+	// For example: "openebs.io/scp-lvm-provisioner" or "openebs.io/scp-device-provisioner".
+	// +optional
+	DefaultStorageProvisioner string `json:"defaultStorageProvisioner,omitempty"`
+}
+
+// StorageCohortStatus stores information about the current status of a storage cohort.
+type StorageCohortStatus struct {
+	// Components is an array of different component's conditions which the cohort is comprised of.
+	// +optional
+	Components []ComponentStatus `json:"components,omitempty"`
+
+	// Capabilities represent capabilities that a cohort consists of
+	// +optional
+	Capabilities Capabilities `json:"capabilities,omitempty"`
+}
+
+// ComponentStatus stores information about the current status of storage cohort's components.
+// Note: For scheduling purpose, the scheduler will only br concerned with the CohortCondition
+// to make scheduling decisions. Other components status can be used for monitoring purposes
+// or troubleshooting purpose.
+type ComponentStatus struct {
+	// CohortCondition is an array of current observed cohort conditions.
+	// The Cohort is deemed to be fully functional when its Ready and Schedulable
+	// condition types are `true`. All other types status declares a cohort
+	// to be non-functional.
+	// +optional
+	CohortCondition []CohortCondition `json:"cohortCondition,omitempty"`
+
+	// CohortManagerCondition is an array of current observed cohort manager conditions.
+	// +optional
+	CohortManagerCondition []ComponentCondition `json:"cohortManagerCondition,omitempty"`
+
+	// NodeCondition is an array of current observed cohort's individual nodes conditions.
+	// +optional
+	NodeCondition []CohortNodeCondition `json:"nodeCondition,omitempty"`
```

Comment:
are we saying there will be single entry in `[]CohortNodeCondition` for every node? If yes, can we make it map[string]CohortNodeCondition to make the intend clear. If no, any rationale behind making `Condition []ComponentCondition` as list in `CohortNodeCondition` ? In latter case, we can simplify it with single `Condition ComponentCondition` in CohortNodeCondition rather than using it as list.

Reply:
There will be an array of conditions for each node present in the cohort. Similar to what we have for k8s single node except here we have for every node. Hence, condition array is present inside the NodeCondition array. For better intend, we can make a `map[string][]ComponentCondition`.

Reply:
map[string][]ComponentCondition looks better !!

Reply:
Tried converting into a map but the controller-tool(i.e controller-gen) used to generate CRD yaml files deosn't support map of custom type arrays. It throws the following error: `map values must be a named type, not *ast.ArrayType`.